File: code/SRM 443/CirclesCountry.py
# -*- coding: utf-8 -*-
import math,string,itertools,fractions,heapq,collections,re,array,bisect

class CirclesCountry:
    def leastBorders(self, X, Y, R, x1, y1, x2, y2):
        borders = 0
        for x, y, r in zip(X, Y, R):
            if (x-r < x1 < x+r and  y-r < y1 < y+r) and not (x-r < x2 < x+r and  y-r < y2 < y+r):
                borders += 1
            elif (x-r < x2 < x+r and  y-r < y2 < y+r) and not (x-r < x1 < x+r and  y-r < y1 < y+r):
                borders += 1
            else:
                pass

        return borders

# CUT begin
# TEST CODE FOR PYTHON {{{
import sys, time, math

def tc_equal(expected, received):
    try:
        _t = type(expected)
        received = _t(received)
        if _t == list or _t == tuple:
            if len(expected) != len(received): return False
            return all(tc_equal(e, r) for (e, r) in zip(expected, received))
        elif _t == float:
            eps = 1e-9
            d = abs(received - expected)
            return not math.isnan(received) and not math.isnan(expected) and d <= eps * max(1.0, abs(expected))
        else:
            return expected == received
    except:
        return False

def pretty_str(x):
    if type(x) == str:
        return '"%s"' % x
    elif type(x) == tuple:
        return '(%s)' % (','.join( (pretty_str(y) for y in x) ) )
    else:
        return str(x)

def do_test(X, Y, R, x1, y1, x2, y2, __expected):
    startTime = time.time()
    instance = CirclesCountry()
    exception = None
    try:
        __result = instance.leastBorders(X, Y, R, x1, y1, x2, y2);
    except:
        import traceback
        exception = traceback.format_exc()
    elapsed = time.time() - startTime   # in sec

    if exception is not None:
        sys.stdout.write("RUNTIME ERROR: \n")
        sys.stdout.write(exception + "\n")
        return 0

    if tc_equal(__expected, __result):
        sys.stdout.write("PASSED! " + ("(%.3f seconds)" % elapsed) + "\n")
        return 1
    else:
        sys.stdout.write("FAILED! " + ("(%.3f seconds)" % elapsed) + "\n")
        sys.stdout.write("           Expected: " + pretty_str(__expected) + "\n")
        sys.stdout.write("           Received: " + pretty_str(__result) + "\n")
        return 0

def run_tests():
    sys.stdout.write("CirclesCountry (300 Points)\n\n")

    passed = cases = 0
    case_set = set()
    for arg in sys.argv[1:]:
        case_set.add(int(arg))

    with open("CirclesCountry.sample", "r") as f:
        while True:
            label = f.readline()
            if not label.startswith("--"): break

            X = []
            for i in range(0, int(f.readline())):
                X.append(int(f.readline().rstrip()))
            X = tuple(X)
            Y = []
            for i in range(0, int(f.readline())):
                Y.append(int(f.readline().rstrip()))
            Y = tuple(Y)
            R = []
            for i in range(0, int(f.readline())):
                R.append(int(f.readline().rstrip()))
            R = tuple(R)
            x1 = int(f.readline().rstrip())
            y1 = int(f.readline().rstrip())
            x2 = int(f.readline().rstrip())
            y2 = int(f.readline().rstrip())
            f.readline()
            __answer = int(f.readline().rstrip())

            cases += 1
            if len(case_set) > 0 and (cases - 1) in case_set: continue
            sys.stdout.write("  Testcase #%d ... " % (cases - 1))
            passed += do_test(X, Y, R, x1, y1, x2, y2, __answer)

    sys.stdout.write("\nPassed : %d / %d cases\n" % (passed, cases))

    T = time.time() - 1489167225
    PT, TT = (T / 60.0, 75.0)
    points = 300 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))
    sys.stdout.write("Time   : %d minutes %d secs\n" % (int(T/60), T%60))
    sys.stdout.write("Score  : %.2f points\n" % points)

if __name__ == '__main__':
    run_tests()

# }}}
# CUT end

File: code/SRM 443/CirclesCountry.java
import java.io.*;
import java.util.*;

public class CirclesCountry {
	public int leastBorders(int[] X, int[] Y, int[] R, int x1, int y1, int x2, int y2) {
		return 0;
	}

// CUT begin
	public static void main(String[] args){
		System.err.println("CirclesCountry (300 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            int[] X = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < X.length; ++i)
                X[i] = Integer.parseInt(Reader.nextLine());
            int[] Y = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < Y.length; ++i)
                Y[i] = Integer.parseInt(Reader.nextLine());
            int[] R = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < R.length; ++i)
                R[i] = Integer.parseInt(Reader.nextLine());
            int x1 = Integer.parseInt(Reader.nextLine());
            int y1 = Integer.parseInt(Reader.nextLine());
            int x2 = Integer.parseInt(Reader.nextLine());
            int y2 = Integer.parseInt(Reader.nextLine());
            Reader.nextLine();
            int __answer = Integer.parseInt(Reader.nextLine());

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(X, Y, R, x1, y1, x2, y2, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1489166972;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 300 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(int[] X, int[] Y, int[] R, int x1, int y1, int x2, int y2, int __expected) {
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		CirclesCountry instance = new CirclesCountry();
		int __result = 0;
		try {
			__result = instance.leastBorders(X, Y, R, x1, y1, x2, y2);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__result == __expected) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + __expected);
			System.err.println("           Received: " + __result);
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "CirclesCountry.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}
